runtime/interpreter/interpreter_statement.go

```diff
@@ -338,28 +338,36 @@ func (interpreter *Interpreter) VisitForStatement(statement *ast.ForStatement) S
 		panic(errors.NewUnreachableError())
 	}
 
-	iterator := iterable.Iterator(interpreter)
+	forStmtTypes := interpreter.Program.Elaboration.ForStatementType(statement)
 
 	var index IntValue
 	if statement.Index != nil {
 		index = NewIntValueFromInt64(interpreter, 0)
 	}
 
-	for {
-		value := iterator.Next(interpreter)
-		if value == nil {
-			return nil
-		}
-
+	executeBody := func(value Value) (resume bool) {
 		statementResult, done := interpreter.visitForStatementBody(statement, index, value)
 		if done {
-			return statementResult
+			result = statementResult
 		}
 
+		resume = !done
+
 		if statement.Index != nil {
 			index = index.Plus(interpreter, intOne, locationRange).(IntValue)
 		}
+
+		return
 	}
+
+	iterable.ForEach(
```

Comment:
Switched to `ForEach` as opposed to directly using the iterator here, since we need to track resource references and prevent container mutation while iterating (https://github.com/dapperlabs/cadence-internal/pull/119). 
 
`Array.Iterate(inter, f)` already implements this, so switching to foreach (which delegates to Array.Iterate) allows re-using the same.